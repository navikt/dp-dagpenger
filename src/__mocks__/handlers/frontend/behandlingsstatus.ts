const behandlingsstatusResolver = (req, res, ctx) => {
  return res(
    ctx.json({
      status: "UnderOgFerdigBehandlet",
      antallSÃ¸knader: 3,
      antallVedtak: 2,
    })
  );
};

export default behandlingsstatusResolver;
