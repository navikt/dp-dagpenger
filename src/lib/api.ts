export default function api(endpoint: string): string {
  if (process.env.NODE_ENV == "test") {
    const url = new URL(location.href);
    url.pathname = endpoint;
    return url.toString();
  }
  return `${process.env.NEXT_PUBLIC_BASE_PATH}/api${
    endpoint.startsWith("/") ? "" : "/"
  }${endpoint}`;
}
