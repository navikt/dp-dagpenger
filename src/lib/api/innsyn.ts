import { v4 as uuid } from "uuid";

export async function fetchInnsynAPI(
  token: Promise<string>,
  endpoint: string
): Promise<any> {
  const callId = uuid();

  console.log(`callId (${callId}) - Henter ${endpoint} fra innsyn.`);

  return fetch(`${process.env.INNSYN_API}/${endpoint}`, {
    method: "get",
    headers: { Authorization: `Bearer ${await token}` },
  })
    .then(async (res) => {
      const contentType = res.headers.get("content-type");
      if (!contentType || !contentType.includes("application/json")) {
        console.log(res.status, res.statusText, res.headers);
        throw new TypeError(
          `Fikk ikke JSON fra innsyn. Body: (${await res.text()}).`
        );
      }
      console.log(`callId (${callId}) - Fikk svar fra innsyn`);
      return res;
    })
    .then((data) => data.json())
    .catch((err) =>
      console.error(
        `callId (${callId}) - Kunne ikke hente ${endpoint} fra innsyn. Feilmelding: ${err}`
      )
    );
}
