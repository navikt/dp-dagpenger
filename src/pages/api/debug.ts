import { getSession } from "../../auth/hooks/session";
import { NextApiRequest, NextApiResponse } from "next";
import nc from "next-connect";
import tokenx from "../../auth/middlewares/tokenx";
import { AuthedNextApiRequest } from "../../auth/middlewares";

export default nc()
  .use(tokenx)
  .get(async (req: AuthedNextApiRequest, res: NextApiResponse) => {
    const session = await getSession({ req });
    if (!session) return res.status(401).end();
    const token = await req.getToken(
      session.user.id_token,
      "dev-gcp:teamdagpenger:dp-innsyn"
    );
    const data = await fetch(process.env.INNSYN_API, {
      method: "get",
      headers: { Authorization: `Bearer ${token}` },
    }).then((data) => data.json());

    res.json(data);
  });
