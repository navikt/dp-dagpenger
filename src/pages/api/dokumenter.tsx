import { NextApiResponse } from "next";
import { gql, GraphQLClient } from "graphql-request";
import { v4 as uuidv4 } from "uuid";
import { AuthedNextApiRequest, withMiddleware } from "../../auth/middlewares";
import { Query } from "../../saf";

const endpoint = "https://safselvbetjening.dev-fss-pub.nais.io/graphql";

export type Dokument = {
  journalpostId: string;
  tittel: string;
};

async function hentDokumenter(
  token: string,
  fnr: string
): Promise<Pick<Query, "dokumentoversiktSelvbetjening">> {
  const callId = uuidv4();
  const variables = { fnr };

  const query = gql`
    query dokumentoversiktSelvbetjening($fnr: String!) {
      dokumentoversiktSelvbetjening(ident: $fnr, tema: [DAG, OPP]) {
        tema {
          journalposter {
            journalpostId
            tittel
            journalposttype
            journalstatus
            dokumenter {
              dokumentInfoId
              tittel
            }
          }
        }
      }
    }
  `;

  const client = new GraphQLClient(endpoint, {
    headers: {
      Authorization: `Bearer ${token}`,
      "Nav-Callid": callId,
      "Nav-Consumer-Id": "dp-dagpenger",
    },
  });

  try {
    console.log(`Henter dokumenter med call-id: ${callId}`);
    return await client.request(query, variables);
  } catch (error) {
    console.error(`Feil fra SAF med call-id ${callId}: ${error}`);
    throw error;
  }
}

export async function handleDokumenter(
  req: AuthedNextApiRequest,
  res: NextApiResponse<Dokument[]>
) {
  const user = req.user;
  if (!user) return res.status(401).end();
  const token = await user.tokenFor(
    "dev-fss:teamdokumenthandtering:safselvbetjening"
  );

  let journalposter;
  try {
    const {
      dokumentoversiktSelvbetjening: { tema },
    } = await hentDokumenter(token, user.fnr);
    journalposter = tema[0].journalposter;
  } catch (errors) {
    return res.status(500).send(errors);
  }

  const dokumenter = journalposter.map(({ journalpostId, tittel }) => ({
    id: journalpostId,
    tittel,
  }));

  res.status(200).json(dokumenter);
}

export default withMiddleware(handleDokumenter);
