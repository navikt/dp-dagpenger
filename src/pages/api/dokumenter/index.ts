import { NextApiHandler } from "next";
import { AvsenderMottaker, Datotype, Journalposttype } from "../../../saf";
import { hentDokumentOversikt } from "../../../lib/saf.service";
import { withSentry } from "@sentry/nextjs";
import { getSession } from "../../../lib/auth.utils";
import { decodeJwt } from "@navikt/dp-auth";
import { safAudience } from "../../../lib/audience";
import { logger } from "@navikt/next-logger";

export type Journalpost = {
  journalpostId: string;
  journalposttype: Journalposttype;
  tittel: string;
  dato: string;
  tema: string;
  dokumenter: Dokument[];
  avsender: AvsenderMottaker;
  mottaker: AvsenderMottaker;
  brukerErAvsenderMottaker: boolean;
};
export type Dokument = {
  id: string;
  tittel: string;
  links: Link[];
  type: DokumentType;
  brukerHarTilgang: boolean;
};
export type DokumentType = "Hoved" | "Vedlegg";
export type Link = { href: string; rel: LinkRel; type: LinkType };
export type LinkType = "GET" | "POST";
export type LinkRel = "preview";

export const handleDokumenter: NextApiHandler<Journalpost[]> = async (
  req,
  res
) => {
  const session = await getSession(req);
  if (!session) return res.status(401).end();
  const payload = decodeJwt(session.token);

  const fnr = payload.pid as string;

  let jposter;
  try {
    const {
      dokumentoversiktSelvbetjening: { journalposter },
    } = await hentDokumentOversikt(await session.apiToken(safAudience), fnr);
    jposter = journalposter;

    logger.info(`Hentet ${jposter.length} journalposter`);
  } catch (errors) {
    logger.error(`Feil fra SAF: ${errors.response}`);
    return res.status(500).end();
  }

  const mapTilRettDato = ({ relevanteDatoer, ...rest }) => {
    const { dato } = relevanteDatoer.find(
      (dato) => dato.datotype == Datotype.DatoOpprettet
    );
    return {
      dato,
      ...rest,
    };
  };

  const berikAvsenderMottaker = ({ avsender, mottaker, ...rest }) => {
    const brukerEr = (am: AvsenderMottaker) =>
      am.type == "FNR" && am.id === fnr;

    const brukerErAvsenderEllerMottaker = () => {
      if (avsender) return brukerEr(avsender);
      if (mottaker) return brukerEr(mottaker);
      return false;
    };
    return {
      brukerErAvsenderMottaker: brukerErAvsenderEllerMottaker(),
      ...rest,
    };
  };

  const journalpostRespons: Journalpost[] = jposter
    .map(mapTilRettDato)
    .map(berikAvsenderMottaker)
    .map(({ journalpostId, dokumenter, ...rest }) => {
      const berikDokmedType = (dok, index) => ({
        // FÃ¸rste vedlegg er alltid hoveddokument
        type: index == 0 ? "Hoved" : "Vedlegg",
        ...dok,
      });

      const berikMedBrukerTilgang = ({ dokumentvarianter, ...rest }) => {
        const erArkiv = (variant) => variant.variantformat === "ARKIV";
        const getArkivVariant = (dokVarianter) => {
          if (dokVarianter) return dokVarianter.find(erArkiv);
          return null;
        };

        const dokumentetKanVises = () => {
          const variant = getArkivVariant(dokumentvarianter);
          if (variant) return variant.brukerHarTilgang;
          return false;
        };
        return { brukerHarTilgang: dokumentetKanVises(), ...rest };
      };

      const byggForhandsvisningLink = ({ dokumentInfoId, ...rest }) => ({
        links: [
          {
            href: `${process.env.NEXT_PUBLIC_BASE_PATH}/api/dokumenter/${journalpostId}/${dokumentInfoId}/forhandsvisning`,
            rel: "preview",
            type: "GET",
          },
        ],
        dokumentInfoId,
        ...rest,
      });

      return {
        journalpostId,
        ...rest,
        dokumenter: dokumenter
          .map(berikDokmedType)
          .map(berikMedBrukerTilgang)
          .map(byggForhandsvisningLink)
          .map(({ dokumentInfoId, ...rest }) => ({
            id: dokumentInfoId,
            ...rest,
          })),
      };
    });

  res.json(journalpostRespons);
};

export default withSentry(handleDokumenter);
