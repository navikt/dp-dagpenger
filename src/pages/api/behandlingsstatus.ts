import { NextApiHandler } from "next";
import { withSentry } from "@sentry/nextjs";
import { getSession } from "../../lib/auth.utils";
import { fetchInnsynAPI } from "../../lib/api/innsyn";

const antallDager = 28;

export type Behandlingsstatus = {
  status: Status;
  antallSøknader: number;
  antallVedtak: number;
};

export type Status =
  | null
  | "UnderBehandling"
  | "FerdigBehandlet"
  | "UnderOgFerdigBehandlet";

export async function hentBehandlingsstatus(
  token: Promise<string>
): Promise<Behandlingsstatus> {
  const fom = new Date();
  fom.setDate(fom.getDate() - antallDager);

  const [søknader, vedtak] = await Promise.all([
    fetchInnsynAPI(token, `soknad?soktFom=${getISODate(fom)}`),
    fetchInnsynAPI(token, `vedtak?fattetFom=${getISODate(fom)}`),
  ]);

  if (!søknader || !vedtak) {
    // Kunne ikke hente søknader og/eller vedtak, så vi viser ikke status
    return { status: null, antallSøknader: 0, antallVedtak: 0 };
  }

  const antallSøknader = søknader.length;
  const antallVedtak = vedtak.length;

  let status: Status = null;
  if (antallSøknader > antallVedtak && antallVedtak > 0) {
    status = "UnderOgFerdigBehandlet";
  } else if (antallVedtak > 0) {
    status = "FerdigBehandlet";
  } else if (antallSøknader > 0) {
    status = "UnderBehandling";
  }

  return {
    status,
    antallSøknader,
    antallVedtak,
  };
}

export const handleBehandlingsstatus: NextApiHandler<Behandlingsstatus> =
  async (req, res) => {
    const { token, apiToken } = await getSession(req);
    if (!token) return res.status(401).end();

    const audience = `${process.env.NAIS_CLUSTER_NAME}:teamdagpenger:dp-innsyn`;

    return hentBehandlingsstatus(apiToken(audience)).then(res.json);
  };

export default withSentry(handleBehandlingsstatus);

function getISODate(date: Date) {
  return (
    date.getFullYear() +
    "-" +
    ("0" + (date.getMonth() + 1)).slice(-2) +
    "-" +
    ("0" + date.getDate()).slice(-2)
  );
}
