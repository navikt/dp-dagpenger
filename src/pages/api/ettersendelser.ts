import { NextApiHandler } from "next";
import { withSentry } from "@sentry/nextjs";
import { getSession } from "../../lib/auth.utils";
import { fetchInnsynAPI } from "../../lib/api/innsyn";
import { innsynAudience } from "../../lib/audience";

export interface EttersendingResultat {
  results: Ettersending[];
  successFullSources: string[];
  failedSources: string[];
}

export interface Ettersending {
  tittel?: string;
  s√∏knadId?: string;
  datoInnsendt: string;
}

export async function hentEttersendelser(token: Promise<string>): Promise<any> {
  return fetchInnsynAPI(token, `ettersendelser`);
}

export const handleEttersendelser: NextApiHandler<EttersendingResultat> =
  async (req, res) => {
    const { token, apiToken } = await getSession(req);
    if (!token) return res.status(401).end();

    const audience = `${process.env.NAIS_CLUSTER_NAME}:teamdagpenger:dp-innsyn`;
    return hentEttersendelser(apiToken(audience)).then(res.json);
  };

export default withSentry(handleEttersendelser);
