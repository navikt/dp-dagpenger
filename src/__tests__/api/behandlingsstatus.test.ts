import { createMocks } from "node-mocks-http";
import { handleBehandlingsstatus } from "../../pages/api/behandlingsstatus";
import { server } from "../../../jest.setup";
import { rest } from "msw";

describe("/api/behandlingsstatus", () => {
  test("svarer med behandlingsstatus null uten vedtak og søknad", async () => {
    med({
      antallSøknader: 0,
      antallVedtak: 0,
    });

    const res = await hentBehandlingsstatus();

    expect(res._getStatusCode()).toBe(200);
    expect(res._getJSONData()["status"]).toBeNull();
    expect(res._getData()).toMatchSnapshot();
  });

  test("svarer med behandlingsstatus UnderBehandling med 1 søknad og 0 vedtak", async () => {
    med({
      antallSøknader: 1,
      antallVedtak: 0,
    });

    const res = await hentBehandlingsstatus();

    expect(res._getStatusCode()).toBe(200);
    expect(res._getJSONData()["status"]).toEqual("UnderBehandling");
    expect(res._getData()).toMatchSnapshot();
  });

  test("svarer med behandlingsstatus UnderOgFerdigBehandlet med 1 søknad og 1 vedtak", async () => {
    med({
      antallSøknader: 1,
      antallVedtak: 1,
    });

    const res = await hentBehandlingsstatus();

    expect(res._getStatusCode()).toBe(200);
    expect(res._getJSONData()["status"]).toEqual("UnderOgFerdigBehandlet");
    expect(res._getData()).toMatchSnapshot();
  });
  test("svarer med behandlingsstatus UnderOgFerdigBehandlet med 2 søknad og 1 vedtak", async () => {
    med({
      antallSøknader: 2,
      antallVedtak: 1,
    });

    const res = await hentBehandlingsstatus();

    expect(res._getStatusCode()).toBe(200);
    expect(res._getJSONData()["status"]).toEqual("UnderOgFerdigBehandlet");
    expect(res._getData()).toMatchSnapshot();
  });

  function med({
    antallSøknader,
    antallVedtak,
  }: {
    antallSøknader: number;
    antallVedtak: number;
  }) {
    server.use(
      rest.get("http://dp-innsyn/soknad", (req, res, ctx) => {
        return res(ctx.json(new Array(antallSøknader)));
      })
    );
    server.use(
      rest.get("http://dp-innsyn/vedtak", (req, res, ctx) => {
        return res(ctx.json(new Array(antallVedtak)));
      })
    );
  }

  async function hentBehandlingsstatus() {
    const { req, res } = createMocks({
      method: "GET",
      user: {
        fnr: "123123123",
        tokenset: {
          access_token: "123",
        },
        tokenFor: (token) => token,
      },
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await handleBehandlingsstatus(req, res);
    return res;
  }
});
