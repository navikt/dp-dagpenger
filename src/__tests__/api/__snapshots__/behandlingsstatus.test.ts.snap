// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/behandlingsstatus med manglende body 1`] = `"{\\"status\\":null,\\"antallSøknader\\":0,\\"antallVedtak\\":0}"`;

exports[`/api/behandlingsstatus svarer med behandlingsstatus FerdigBehandlet med 0 søknad og 1 vedtak 1`] = `"{\\"status\\":\\"FerdigBehandlet\\",\\"antallSøknader\\":0,\\"antallVedtak\\":1}"`;

exports[`/api/behandlingsstatus svarer med behandlingsstatus FerdigBehandlet med 1 søknad og 1 vedtak 1`] = `"{\\"status\\":\\"FerdigBehandlet\\",\\"antallSøknader\\":1,\\"antallVedtak\\":1}"`;

exports[`/api/behandlingsstatus svarer med behandlingsstatus FerdigBehandlet med 1 søknad og 2 vedtak 1`] = `"{\\"status\\":\\"FerdigBehandlet\\",\\"antallSøknader\\":1,\\"antallVedtak\\":2}"`;

exports[`/api/behandlingsstatus svarer med behandlingsstatus UnderBehandling med 1 søknad og 0 vedtak 1`] = `"{\\"status\\":\\"UnderBehandling\\",\\"antallSøknader\\":1,\\"antallVedtak\\":0}"`;

exports[`/api/behandlingsstatus svarer med behandlingsstatus UnderOgFerdigBehandlet med 2 søknad og 1 vedtak 1`] = `"{\\"status\\":\\"UnderOgFerdigBehandlet\\",\\"antallSøknader\\":2,\\"antallVedtak\\":1}"`;

exports[`/api/behandlingsstatus svarer med behandlingsstatus null uten vedtak og søknad 1`] = `"{\\"status\\":null,\\"antallSøknader\\":0,\\"antallVedtak\\":0}"`;
